src/compute/src/arrangement/manager.rs

```diff
@@ -170,6 +178,128 @@ impl SpecializedTraceHandle {
+///
+/// A padded trace contains empty data for all times greater than or equal to its `padded_since`
+/// and less than the logical compaction frontier of the inner `trace`.
+#[derive(Clone)]
+pub struct PaddedTrace<Tr> {
+    /// The wrapped trace.
+    trace: Tr,
+    /// The frontier from which the trace is padded, or `None` if it is not padded.
+    padded_since: Option<Antichain<Timestamp>>,
+}
+
+impl<Tr> From<Tr> for PaddedTrace<Tr> {
+    fn from(trace: Tr) -> Self {
+        Self {
+            trace,
+            padded_since: None,
+        }
+    }
+}
+
+impl<Tr> PaddedTrace<Tr> {
+    /// Turns this trace into a padded version that reports empty data for all times less than the
+    /// trace's current logical compaction frontier.
+    fn into_padded(mut self) -> Self {
+        self.padded_since = Some(Antichain::from_elem(Timestamp::MIN));
+        self
+    }
+}
+
+impl<Tr> TraceReader for PaddedTrace<Tr>
+where
+    Tr: TraceReader<Time = Timestamp>,
+{
+    type Key<'a> = Tr::Key<'a>;
+    type Val<'a> = Tr::Val<'a>;
+    type Time = Tr::Time;
+    type TimeGat<'a> = Tr::TimeGat<'a>;
+    type Diff = Tr::Diff;
+    type DiffGat<'a> = Tr::DiffGat<'a>;
+    type Batch = Tr::Batch;
+    type Storage = Tr::Storage;
+    type Cursor = Tr::Cursor;
+
+    fn cursor_through(
+        &mut self,
+        upper: AntichainRef<Self::Time>,
+    ) -> Option<(Self::Cursor, Self::Storage)> {
+        self.trace.cursor_through(upper)
+    }
+
+    fn set_logical_compaction(&mut self, frontier: AntichainRef<Self::Time>) {
+        let Some(padded_since) = &mut self.padded_since else {
+            self.trace.set_logical_compaction(frontier);
+            return;
+        };
+
+        // If a padded trace is compacted to some frontier less than the inner trace's compaction
+        // frontier, advance the `padded_since`. Otherwise discard the padding and apply the
+        // compaction to the inner trace instead.
+        let trace_since = self.trace.get_logical_compaction();
+        if PartialOrder::less_than(&frontier, &trace_since) {
+            if PartialOrder::less_than(&padded_since.borrow(), &frontier) {
+                *padded_since = frontier.to_owned();
+            }
+        } else {
+            self.padded_since = None;
+            self.trace.set_logical_compaction(frontier);
+        }
+    }
+
+    fn get_logical_compaction(&mut self) -> AntichainRef<Self::Time> {
+        match &self.padded_since {
+            Some(since) => since.borrow(),
+            None => self.trace.get_logical_compaction(),
+        }
+    }
+
+    fn set_physical_compaction(&mut self, frontier: AntichainRef<Self::Time>) {
+        self.trace.set_physical_compaction(frontier);
+    }
+
+    fn get_physical_compaction(&mut self) -> AntichainRef<Self::Time> {
+        self.trace.get_logical_compaction()
+    }
+
+    fn map_batches<F: FnMut(&Self::Batch)>(&self, f: F) {
+        // TODO: need to insert empty batch?
```

Comment:
I think the answer is "yes", but I think we need to get @frankmcsherry's take on this. The `map_batches` API isn't spec'ed out enough to tell whether that's a correct implementation or no.

Reply:
`map_batches` should present a sequence of batches that are complete, i.e. cover all ranges of times from the start of time up until the currentmost frontier you want to report.

Reply:
From discussion with @antiguru: There is no need (and it would be wrong) to insert an additional batch here because calling `map_batches` on the inner trace already produces a batch starting from the beginning of time.